scripts/template_upload_gce.py

```diff
@@ -0,0 +1,160 @@
+            return (file_name, file_path)
+        os.remove(file_name)
+    print "Downloading: %s Bytes: %s" % (file_name, file_size)
+    f = open(file_name, 'wb')
+    os.system('cls')
+    file_size_dl = 0
+    block_sz = 8192
+    while True:
+        buffer = u.read(block_sz)
+        if not buffer:
+            break
+
+        file_size_dl += len(buffer)
+        f.write(buffer)
+        status = r"%10d  [%3.2f%%]" % (file_size_dl, file_size_dl * 100. / file_size)
+        status = status + chr(8) * (len(status) + 1)
+        print status,
+
+    f.close()
+    return (file_name, file_path)
+
+
+def check_template_name(name):
+    pattern = re.compile("(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)")
+    match = pattern.match(name)
+    if isinstance(match.group(), str) and match.group() != name:
+        name = cfme_data['basic_info']['appliance_template']
+    return name
+
+
+def upload_template(project, zone, service_account, image_url,
+        template_name, bucket_name, provider):
+    try:
+        print("GCE:{} Starting image downloading {} ...".format(provider, image_url))
+        file_name, file_path = download_image_file(image_url)
+        print("GCE:{} Image downloaded {} ...".format(provider, file_path))
+
+        print("GCE:{} Creating bucket and uploading file {}...".format(provider, bucket_name))
+        gcloud = GoogleCloudSystem(project=project, zone=zone, service_account=service_account)
+        gcloud.create_bucket(bucket_name)
+        blob_name = gcloud.get_file_from_bucket(bucket_name, file_name)
+        if not blob_name:
+            res, blob_name = gcloud.upload_file_to_bucket(bucket_name, file_path)
+        print("GCE:{} File uploading done ...".format(provider))
+
+        print("GCE:{} Creating template/image {}...".format(provider, template_name))
+        template_name = check_template_name(template_name)
+        gcloud.create_image(image_name=template_name, bucket_url=blob_name.get('selfLink'))
+        print("GCE:{} Successfully uploaded the template.".format(provider))
+    except Exception as e:
+        print(e)
+        print("GCE:{} Error occurred in upload_template".format(provider, template_name))
+        return False
+    finally:
+        print("GCE:{} End template {} upload...".format(provider, template_name))
+
+
+def run(**kwargs):
+    thread_queue = []
+    for provider in list_providers("gce"):
+        mgmt_sys = cfme_data['management_systems'][provider]
+        gce_credentials = credentials[mgmt_sys['credentials']]
+
+        service_account = gce_credentials['service_account']
+        project = mgmt_sys['project']
+        zone = mgmt_sys['zone']
+        thread = Thread(target=upload_template,
+                        args=(project, zone, service_account, kwargs.get('image_url'),
+                            kwargs.get('template_name'), kwargs.get('bucket_name'), provider))
+        thread.daemon = True
+        thread_queue.append(thread)
```

Comment:
is it ok to just start as many threads as there are gce providers, (i suppose we dont have that many in a call)?